lib/puppet/type/user.rb

```diff
@@ -459,40 +467,47 @@ def insync?(current)
       end
     end
 
-    # Autorequire the group, if it's around
+    # Autorequire groups, if they're around
     autorequire(:group) do
       autos = []
 
-      # autorequire primary group, if managed
-      obj = @parameters[:gid]
-      groups = obj.shouldorig if obj
-      if groups
-        groups = groups.collect { |group|
-          if group.is_a?(String) && group =~/^\d+$/
-            Integer(group)
-          else
-            group
-          end
-        }
-        groups.each { |group|
-          case group
-          when Integer
-            resource = catalog.resources.find { |r| r.is_a?(Puppet::Type.type(:group)) && r.should(:gid) == group }
-            if resource
-              autos << resource
+      if self[:ensure] != :absent
+        if @parameters[:gid]&.shouldorig
+          groups_in_catalog = catalog.resources.filter { |r| r.is_a?(Puppet::Type.type(:group)) }
+          autos += @parameters[:gid].shouldorig.filter_map do |group|
+            if (group.is_a?(String) && group.match?(/^\d+$/)) || group.is_a?(Integer)
+              gid = Integer(group)
+              groups_in_catalog.find { |r| r.should(:gid) == gid }
+            else
+              group
             end
-          else
-            autos << group
           end
-        }
+        end
+
+        if @parameters[:groups]&.should
+          autos += @parameters[:groups].should.split(",")
+        end
       end
 
-      # autorequire groups, excluding primary group, if managed
-      obj = @parameters[:groups]
-      if obj
-        groups = obj.should
-        if groups
-          autos += groups.split(",")
+      autos
+    end
+
+    # Autobefore the primary group, if it's around. Otherwise group removal
+    # fails when that group is also ensured absent.
+    autobefore(:group) do
```

Comment:
There's a long standing bug in puppet that auto* relationships and ensure => absent don't work well together. I filed a redmine ticket a long time ago, which was exported to https://puppet.atlassian.net/browse/PUP-2451. Ideally I think puppet would know to reverse the direction for any auto* relationship when ensuring absent. Worse, if you try to set an explicit relationship, it will create a cycle due to the auto relationship.

Reply:
I have written Puppet code where I had a lot of conditionals based on absent/present. I'd love for Puppet to be smarter here. But reading https://github.com/voxpupuli/puppet-systemd/blob/dbaa56f690e36f9f164fee464bf7b0e8a9a6d37c/manifests/unit_file.pp#L131-L142 I really don't know how it could be designed in a way that's not more confusing than the status quo.

Overall few people write modules with (full) support to ensure things are absent.